gprofiler/utils/node.py

```diff
@@ -0,0 +1,176 @@
+import json
+import os
+import shutil
+import signal
+import stat
+from typing import Dict, List
+
+import psutil
+import requests
+from granulate_utils.linux.ns import get_proc_root_path, get_process_nspid, resolve_proc_root_links, run_in_ns
+from granulate_utils.linux.process import is_musl
+from retry import retry
+from websocket import create_connection
+from websocket._core import WebSocket
+
+from gprofiler.log import get_logger_adapter
+from gprofiler.metadata import get_exe_version
+from gprofiler.utils import add_permission_dir, pgrep_exe, resource_path
+
+logger = get_logger_adapter(__name__)
+
+DSO_GIT_REVISION = "20eb88a"
+
+
+class NodeDebuggerUrlNotFound(Exception):
+    pass
+
+
+class NodeDebuggerUnexpectedResponse(Exception):
+    pass
+
+
+def get_dest_inside_container(musl: bool, node_version: str) -> str:
+    libc = "musl" if musl else "glibc"
+    return os.path.join("/", "tmp", "node_module", DSO_GIT_REVISION, libc, node_version)
+
+
+def _start_debugger(pid: int) -> None:
+    # for windows: in shell node -e "process._debugProcess(PID)"
+    os.kill(pid, signal.SIGUSR1)
+
+
+@retry(NodeDebuggerUrlNotFound, 5, 1)
+def _get_debugger_url() -> str:
+    # when killing process with SIGUSR1 it will open new debugger session on port 9229,
+    # so it will always the same
+    port = 9229
```

Comment:
By the way - if the port is already caught by another process, what will happen? No debug port will be opened?

We possibly need to add special handling for that here. Can you ensure the server you're talking with is actually the PID we desire? What if it already has the module loaded from e.g a previous run of gProfiler?

I think the best structure would be to add sanity checks before invoking the debugger, and after you have a debugger socket, ensure you're working with the desired process. The simplest way I can think of (which isn't very simple) is to compare the NSpid of the process bundled with its `/proc/self/ns/pid` link name (which corresponds to the PID namespace identifier of the process). Together these 2 are a unique identifier for a PID, seen both from the inner NS (process) and the outer NS (gprofiler)

Reply:
(for example, we might attempt to load the DSO of incorrect version to a process)

Reply:
If port is already opened it won't open. We can have one of two solutions for having multiple node processes in same network namespace:

1.  Close debugger after generating perf maps and starting it each time we want to do any operation on debugger. In this method no two debuggers will be open at the same time. This will also be useful when someone started container with i. e. `--network=host`. (I've started implementing that)
2.  We can change debug port, but that can only be done inside process (we need to change `process.debugPort`). This means we need to start it on port 9229 anyway, then change `debugPort` and restart debugger.

In both methods we have to start it at 9229 port and I didn't find any solution for that.

Answering your other questions:
If module is already loaded, CDP request will respond with False when loading module
I'll add this sanity check

Reply:
The 2 options you suggested are viable but I think we'll leave the decision for a future PR. I am meanwhile opening a ticket from it.

Reply:
We have 2 issues remaining here:
1. How do we know we're talking with the process we think we're talking with (nevermind if another process in the namespace could not use the port - we just need to know we're talking not with the expected one).
2. How do we check if the module is already loaded (and ideally, if possible, allow to reload once we change version in new gprofiler release, as we do for async-profiler).

Reply:
1. I've already added logic that checks if `/proc/<pid>/ns/pid` in gprofiler ns is equal to `/proc/<nspid>/ns/pid` inside given namespace. Tried to do it with `/proc/self` but i was getting no such file or directory. This should be enough to check if we are in same namespace. If we want to also check if it is the same process in given namespace we can query debugger for it's pid (i.e. reading `process.pid`) and compare it with expected pid inside namespace.

2. Start function of module returns true if module is loaded and false if it isn't. We can possibly utilize that. Second option is to fork that  module and add our own method, shouldn't be that hard since it is only checking, if handler is nullptr. [ref](https://github.com/mmarchini/node-linux-perf/blob/master/linux-perf.cc#L41)

Reply:
1. Sounds good - we compare both the `ns/pid` and the NSPID, if both match it's the same process.
2. Cases that we want to check: first: already running module of the same version - we detect it and log it. second: upon upgrade of linux-perf, gprofiler uses the new version of linux-perf (2 are loaded in the memory space)

Reply:
2. was chcked - we're good

Reply:
1. you implemented a check but I'm not sure it's enough, I'll comment on the new check.